designs/0013-profiling.md

```diff
@@ -0,0 +1,420 @@
+but is legal code:
+
+```stan
+model {
+  profile("model") {
+    sigma ~ normal(0, 1);
+    b ~ normal(0, 1);
+  }
+  profile("model") {
+    y ~ normal(X * b, sigma);
+  }  
+}
+```
+
+## Blocks and loops
+
+As an extended example of how a `profile` statement works with blocks and loops,
+consider the construction of the Cholesky factor of a Gaussian process kernel:
+
+```stan
+transformed parameters {
+  matrix[N, N] L_K;
+  matrix[N, N] K;
+  profile("gp"){
+    K = gp_exp_quad_cov(x, alpha, rho);
+  }
+  real sq_sigma = square(sigma);
+
+  for (n in 1:N) {
+    profile("add diagonal") {
+      K[n, n] = K[n, n] + sq_sigma;
+    }
+  }
+  profile("cholesky") {
+    L_K = cholesky_decompose(K);
+  }  
+}
+```
+
+There are three profiles here. The first profile, `gp`, will time the
+statement `K = gp_exp_quad_cov(x, alpha, rho)`. The second profile,
+`add diagonal`, will measure the cost of each iteration of the for loop 
+and accumulate that under the `add diagonal` profile name.
+
+The third profile, `cholesky`, measures the cost of
+`cholesky_decompose`.
+
+## Nested profiles
+
+Nested profiles can be used when the overall cost and the individual cost
+of operations are interesting. In this case:
+
+```stan
+model {
+  profile("model");
+  {
+    profile("prior") {
+      sigma ~ normal(0, 1);
+      b ~ normal(0, 1);
+    }    
+  }
+  profile("likelihood") {
+    y ~ normal(X * b, sigma);
+  }  
+}
+```
+
+## Different blocks and user-defined functions
+
+`profile` statements are fine to use in user-defined functions, though using
+one in a recursive function will result in a runtime error. In the example here,
+the user-defined functions is used in multiple blocks. Because of this
+the timing results reported for the profile `myfunc` will include timing from
+the model block (that has autodiff and runs for every gradient evaluation) and
+the generated quantities (that does not have autodiff and only runs once
+for every MCMC draw that is printed).
+
+```stan
+functions {
+  real myfunc(vector alpha) {
+    profile("myfunc") {
+      ...;
+    }    
+  }
+}
+...
+model {
+  real x = myfunc(alpha);
+  ...
+}
+...
+generated quantities {
+  real x = myfunc(alpha);
+  ...
+}
+```
+
+There is ambiguity in what the `myfunc` profile is timing. If in a specific
+situation is was important to resolve this ambiguity, then the timing
+statements should be moved from inside the function to where the function
+is used:
+
+```
+model {
+  real x;
+  profile("myfunc - model") {
+    x = myfunc(alpha);
+  }
+  ...
+}
+...
+generated quantities {
+  real x;
+  profile("myfunc - generated quantities") {
+    x = myfunc(alpha);
+  }
+  ...
+}
+```
+
+## Note on parameters block
+
+Because only parameter declaration statements are allowed in the `parameters`
+block, `profile` cannot be used there. This means that the `lower`, `upper`,
+`multiply` and `offset` constraints/transforms can not be profiled.
+
```

Comment:
Does it make sense to say in general we don't profile declaration, no matter which block it is in?

Reply:
One could technically define a local variable inside a profile.

```stan
profile("name) {
    vector a = b +c;
    d = a / 2;
}
```

Maybe "we can not profile top-level declarations".

Reply:
Syntax still allows the following, does it?
```stan
parameters {
profile("xxx") {
real x;
real y;
}
//...
}
```

Reply:
No, because parameters only allows top-leve variable declaration and nothing else.

```stan
parameters{
    profile("yes") {
        real a;
        real b;
    }
    real y;
}
```
errors with  `Expected top-level variable declaration or "}".`

While 

```stan
parameters{
    real y;
    profile("yes") {
        real a;
        real b;
    }
}
```
errors with `Only top-level variable declarations allowed in data and parameters blocks.`